pilot/pkg/xds/endpoint_builder.go

```diff
@@ -292,12 +292,13 @@ func (b *EndpointBuilder) buildLocalityLbEndpointsFromShards(
 	}
 	// The shards are updated independently, now need to filter and merge for this cluster
 	for _, clusterID := range keys {
-		endpoints := shards.Shards[clusterID]
 		// If the downstream service is configured as cluster-local, only include endpoints that
-		// reside in the same cluster.
-		if isClusterLocal && (cluster.ID(clusterID) != b.clusterID) {
+		// reside in the same cluster. A clusterID of "" may indicate this is a ServiceEntry registry and should be
```

Comment:
When would this happen ? Can you add more comments and example. Unlike the grpcgen changes, this is critical path code.

Reply:
See the discussion in https://github.com/istio/istio/pull/34480

The tests hit istiod and we only send cluster local endpoints for that service. The test worked before because we didn't set a proper cluster Id in metadata but now we do.